matchzoo/preprocessors/cdssm_preprocessor.py

```diff
@@ -0,0 +1,104 @@
+"""CDSSM Preprocessor."""
+
+import logging
+
+from tqdm import tqdm
+
+from matchzoo import build_vocab_unit
+from matchzoo import chain_transform
+from matchzoo import DataPack
+from matchzoo import engine
+from matchzoo import processor_units
+
+logger = logging.getLogger(__name__)
+tqdm.pandas()
+
+
+class CDSSMPreprocessor(engine.BasePreprocessor):
+    """CDSSM Model preprocessor."""
+
+    def __init__(self, text_len: int = 10, with_word_hashing=True):
+        """
+        CDSSM Model preprocessor.
+
+        The word hashing step could eats up a lot of memory. To workaround
+        this problem, set `with_word_hashing` to `False` and use a
+        :class:`matchzoo.DynamicDataGenerator` with a
+        :class:`matchzoo.processor_units.WordHashingUnit`.
+
+        :param text_len: Fixed text length, cut original text if it is longer
+         or pad if shorter.
+        :param with_word_hashing: Include a word hashing step if `True`.
+
+        Example:
+            >>> import matchzoo as mz
+            >>> train_data = mz.datasets.toy.load_data()
+            >>> test_data = mz.datasets.toy.load_data(stage='test')
+            >>> cdssm_preprocessor = mz.preprocessors.CDSSMPreprocessor()
+            >>> train_data_processed = cdssm_preprocessor.fit_transform(
+            ...     train_data
+            ... )
+            >>> type(train_data_processed)
+            <class 'matchzoo.data_pack.data_pack.DataPack'>
+            >>> test_data_transformed = cdssm_preprocessor.transform(test_data)
+            >>> type(test_data_transformed)
+            <class 'matchzoo.data_pack.data_pack.DataPack'>
+
+        """
+        super().__init__()
+        self._text_len = text_len
+        self._with_word_hashing = with_word_hashing
+
+    def fit(self, data_pack: DataPack, verbose=1):
+        """
+        Fit pre-processing context for transformation.
+
+        :param verbose: Verbosity.
+        :param data_pack: Data_pack to be preprocessed.
+        :return: class:`CDSSMPreprocessor` instance.
+        """
+        units = self._default_processor_units()
+        units.append(processor_units.NgramLetterUnit())
+        data_pack = data_pack.apply_on_text(chain_transform(units),
+                                            verbose=verbose)
+        vocab_unit = build_vocab_unit(data_pack, verbose=verbose)
+
+        self._context['vocab_unit'] = vocab_unit
+        num_letter = len(vocab_unit.state['term_index']) + 1
+        self._context['input_shapes'] = [(self._text_len, num_letter),
+                                         (self._text_len, num_letter)]
```

Comment:
What's the point of this duplication? i.e. why not just use `self._context['input_shapes'] = [self._text_len, num_letter]`? Since you customized `Input` in `build`, it's ok to customize your `input_shapes` form as well, as long as these two are consistent.

Reply:
I try to simplify input shape as you recommend, but it is not iterable and can not pass `test_model.py`. I think the duplicate shape is set for different `query` and `document` text length, which not considered in `CDSSM` model. So I reset the input shapes to a list of tuple.

Reply:
Try `self._context['input_shapes'] = [(self._text_len,), (num_letter,)]`. And you may also use `BaseModel._make_inputs()`.

Reply:
I use `BaseModel._make_inputs()` instead and it works!